### PR TITLE
Add the ctranslate command

### PR DESCRIPTION
**Pull request by xpple**
This PR adds a translation command using one of Google Translate's hidden APIs. Omitting the `from` parameter makes Google Translate auto-detect the language. When the `to` parameter is omitted the client's language will be used to translate to.

### CODE REVIEW

File: src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+public class TranslateCommand {
+
+    private static final SimpleCommandExceptionType UNKNOWN_ERROR_EXCEPTION = new SimpleCommandExceptionType(Text.translatable("commands.ctranslate.unknownError"));
+
+    private static final String url = "https://translate.googleapis.com/translate_a/single?client=gtx&dt=t";
+
+    private static final HttpClient httpClient = HttpClient.newHttpClient();
+
+    private static final int DURATION = 5; // seconds
```
Comment (haykam821) [inline src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java]: This constant would be more clear if it was named `TIMEOUT` to match its usage.

File: src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    public static void register(CommandDispatcher<FabricClientCommandSource> dispatcher) {
+        dispatcher.register(literal("ctranslate")
+                .then(argument("query", translationQuery())
+                        .executes(ctx -> translate(ctx.getSource(), getTranslationQuery(ctx, "query")))));
+    }
+
+    private static int translate(FabricClientCommandSource source, TranslationQueryArgumentType.TranslationQuery query) throws CommandSyntaxException {
+        try {
+            HttpRequest request = HttpRequest.newBuilder(createUri(query.from(), query.to(), query.query()))
+                    .timeout(Duration.ofSeconds(DURATION))
```
Comment (haykam821) [inline src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java]: Storing this duration as the timeout constant would make the unit more clear.

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource;
+import net.minecraft.command.CommandSource;
+import net.minecraft.text.Text;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+public class TranslationQueryArgumentType implements ArgumentType<TranslationQueryArgumentType.TranslationQuery> {
```
Comment (haykam821) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java:20]: Does this argument type represent a language as the examples imply? 'Query' is confusing.

**Review by haykam821 — DISMISSED** [review]

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java (hunk 3 above)
Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java:20]: Oops, I forgot to change the examples. They should make it more clear.

File: src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+public class TranslateCommand {
+
+    private static final SimpleCommandExceptionType UNKNOWN_ERROR_EXCEPTION = new SimpleCommandExceptionType(Text.translatable("commands.ctranslate.unknownError"));
+
+    private static final String url = "https://translate.googleapis.com/translate_a/single?client=gtx&dt=t";
+
+    private static final HttpClient httpClient = HttpClient.newHttpClient();
+
+    private static final Duration DURATION = Duration.ofSeconds(5);
```
Comment (haykam821) [inline src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java:33]: Would it make sense to move these fields to a separate class to reduce duplication with the `PlayerInfoCommand` class?

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                } else {
+                    to = language;
+                }
+            }
+            return new TranslationQuery(from, to, query);
+        }
+
+        private Option parseFromTo() throws CommandSyntaxException {
+            int start = reader.getCursor();
+            suggestor = suggestions -> {
```
Comment (haykam821) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java:102]: This doesn't seem right to me.

Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java:102]: In what way?

File: src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java (hunk 4 above)
Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java:33]: I guess? I'll let @Earthcomputer decide.

Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java:33]: I think we should do this when at least three classes use these fields.

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java (hunk 5 above)
Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java:102]: I think this it's fine.

File: src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import java.time.Duration;
+
+import static net.earthcomputer.clientcommands.command.arguments.TranslationQueryArgumentType.*;
+import static net.fabricmc.fabric.api.client.command.v2.ClientCommandManager.*;
+
+public class TranslateCommand {
+
+    private static final SimpleCommandExceptionType UNKNOWN_ERROR_EXCEPTION = new SimpleCommandExceptionType(Text.translatable("commands.ctranslate.unknownError"));
+
+    private static final String url = "https://translate.googleapis.com/translate_a/single?client=gtx&dt=t";
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/TranslateCommand.java]: This should be in caps: `URL`

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.function.Consumer;
+
+public class TranslationQueryArgumentType implements ArgumentType<TranslationQueryArgumentType.TranslationQuery> {
+
+    private static final Collection<String> EXAMPLES = Arrays.asList("\"Voorbeeld\" from nl to en", "\"Non numeranda, sed ponderanda sunt argumenta\" from la", "\"Cogito, ergo sum\"");
+
+    private static final DynamicCommandExceptionType UNKNOWN_LANGUAGE_CODE_EXCEPTION = new DynamicCommandExceptionType(arg -> Text.translatable("commands.ctranslate.unknownLanguageCode", arg));
+    private static final SimpleCommandExceptionType EXPECTED_FROM_TO_EXCEPTION = new SimpleCommandExceptionType(Text.translatable("commands.ctranslate.expectedFromTo"));
+
+    private static final Collection<String> languages;
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/TranslationQueryArgumentType.java]: This should be in caps too as it's immutable

**Review by Earthcomputer — APPROVED** [review]

**Pull request merged**